controller/volume_controller.go

```diff
@@ -771,33 +782,88 @@ func (vc *VolumeController) cleanupFailedToScheduledReplicas(v *longhorn.Volume,
 
 func (vc *VolumeController) cleanupExtraHealthyReplicas(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) (err error) {
 	healthyCount := vc.getHealthyReplicaCount(rs)
+	if healthyCount <= v.Spec.NumberOfReplicas {
+		return nil
+	}
 
-	if healthyCount > v.Spec.NumberOfReplicas {
-		for _, r := range rs {
-			// Clean up eviction requested replica
-			if r.Status.EvictionRequested {
-				if err := vc.deleteReplica(r, rs); err != nil {
-					vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
-						EventReasonFailedEviction,
-						"volume %v failed to evict replica %v",
-						v.Name, r.Name)
-					return err
-				}
-				logrus.Debugf("Evicted replica %v", r.Name)
-				return nil
+	var cleaned bool
+	if cleaned, err = vc.cleanupEvictionRequestedReplicas(v, rs); err != nil || cleaned {
+		return err
+	}
+
+	if cleaned, err = vc.cleanupAutoBalancedReplicas(v, rs); err != nil || cleaned {
```

Comment:
Suggest moving cleanupAutoBalancedReplicas after cleanupDataLocalityReplicas to keep `data locality` respected first.

Reply:
Needs to keep cleanupAutoBalancedReplicas  before to ensure auto-balance best-effort replica list is respected. I've updated to have the replica list go through getPreferredReplicaCandidatesForDeletion so `data locality` will also get respected.

Reply:
Fixed. tested with [test_replica_auto_balance_zone_best_effort_with_data_locality](https://github.com/longhorn/longhorn-tests/pull/609).